lib/ConstPlugin.js

```diff
@@ -221,6 +221,71 @@ class ConstPlugin {
 							}
 						}
 					);
+					parser.hooks.expressionLogicalOperator.tap(
+						"ConstPlugin",
+						expression => {
+							const param = parser.evaluateExpression(expression.left);
+							const bool = param.asBool();
+							if (typeof bool === "boolean") {
+								// Expressions do not hoist.
+								// It is safe to remove the dead branch.
+								//
+								// ------------------------------------------
+								//
+								// Given the following code:
+								//
+								//   falsyExpression() && someExpression();
+								//
+								// the generated code is:
+								//
+								//   falsyExpression() && false; // "false" could be any value here
+								//
+								// ------------------------------------------
+								//
+								// Given the following code:
+								//
+								//   truthyExpression() && someExpression();
+								//
+								// the generated code is:
+								//
+								//   true && someExpression();
+								//
+								// ------------------------------------------
+								//
+								// Given the following code:
+								//
+								//   truthyExpression() || someExpression();
+								//
+								// the generated code is:
+								//
+								//   truthyExpression() || true; // "true" could be any value here
+								//
+								// ------------------------------------------
+								//
+								// Given the following code:
+								//
+								//   falsyExpression() || someExpression();
+								//
+								// the generated code is:
+								//
+								//   false && someExpression();
+								//
+								const keepLeft =
+									(expression.operator === "&&" && !bool) ||
+									(expression.operator === "||" && bool);
```

Comment:
This will break if more LogicalOperators are added in future. Better filter these two operators at the beginning of the plugin.